hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HRegion.java

```diff
@@ -8458,16 +8452,24 @@ public void startRegionOperation(Operation op) throws IOException {
         coprocessorHost.postStartRegionOperation(op);
       }
     } catch (Exception e) {
-      if (isInterruptableOp) {
-        // would be harmless to remove what we didn't add but we know by 'isInterruptableOp'
-        // if we added this thread to regionLockHolders
-        regionLockHolders.remove(thisThread);
-      }
+      lockHolders.remove(thisThread);
       lock.readLock().unlock();
       throw new IOException(e);
     }
   }
 
+  private static boolean shouldLock(Operation op) {
+    // split, merge or compact region doesn't need to check the closing/closed state or lock the
+    // region
+    List<Operation> lockFreeOperations = Arrays.asList(Operation.MERGE_REGION,
```

Comment:
If we're going to keep this refactor, then we don't need to create a new List every time we call this method. This could be kept somewhere as a static list/set. (Maybe compiler already optimizes this, I don't know.)